kotlinx-coroutines-core/jvm/src/internal/ResizableAtomicArray.kt

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2016-2022 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.internal
+
+import kotlinx.coroutines.*
+import java.util.concurrent.atomic.*
+
+/**
+ * Atomic array with lock-free reads and synchronized modifications. It logically has an unbounded size,
+ * is implicitly filled with nulls, and is resized on updates as needed to grow.
+ */
+internal class ResizableAtomicArray<T>(initialLength: Int) {
+    @Volatile
+    private var array = AtomicReferenceArray<T>(initialLength)
+
+    // for debug output
+    public fun currentLength(): Int = array.length()
+
+    public operator fun get(index: Int): T? {
+        assert { index >= 0 }
+        val array = this.array // volatile read
+        return if (index < array.length()) array[index] else null
+    }
+
+    @Synchronized
```

Comment:
`set` is always guarded by `synchronized(workers)`.

It's better to avoid excessive synchronization; alternatively, you can do `@GuardedBy("this")` or `assert { Thread.holdsLock(this) }`

Reply:
The `Thread.holdsLock(this)` call is extremely expensive. I don't feel it is Ok to have it running even inside `assert { ... }` (that can be still turned on in production when assertions are enabled)>

Just removing `@Synchronized` and documenting that it can be called only under lock is error-prone, since that code lives in a different package from the scheduler and could be, one day, used by someone who did not read the docs (inside or outside scheduler). 

On the other hand, an extra nested `@Synchronized` is usually removed by HotSpot after inlining and it seems, that this accessor is a small-enough method to be always inlined. I have not checked if that is the case for this particular code, though.

As a compromise, `operator fun set(...)` can be renamed to `fun setSynchonized(...)`. This will make all its call-sites more verbose, so we can remove `@Synchronize` without any fear of it being error-prone.